File: static/bitboard.js
function laske_mask(x,y) {
    return (BigInt(1) << ((BigInt(x) - BigInt(1)) * BigInt(7) + BigInt(y) - BigInt(1)));
}

function symboliKohdassa(x,y, bitboard){
    mask = laske_mask(x,y);
    if ((mask & bitboard[0]) != 0) return "white";
    if ((mask & bitboard[1]) != 0) return "black";
    return "tyhja";
}

function make_move(x,y, bitboard) {
    return BigInt(bitboard | laske_mask(x,y));
}
    
function kaannaVasemmalle(x,y, board) {
    var kopio = board;
    for (var i = -1; i <= 1; i++) {
        for (var j = -1; j <= 1; j++) {
            var bitti = (laske_mask(BigInt(i)+x, BigInt(j)+y) & board);
            if (bitti != 0) {
                kopio |= laske_mask(-BigInt(j)+x, BigInt(i)+y);
            } else {
                kopio &= (~laske_mask(-BigInt(j)+x, BigInt(i)+y));
            }
        }
    }
    return kopio;
}
    
function kaannaOikealle(x,y, board) {
    var kopio = board;
    for (var i = -1; i <= 1; i++) {
        for (var j = -1; j <= 1; j++) {
            var bitti = (laske_mask(BigInt(i)+x, BigInt(j)+y) & board);
            if (bitti != 0) {
                kopio |= laske_mask(BigInt(j)+x, -BigInt(i)+y);
            } else {
                kopio &= (~laske_mask(BigInt(j)+x, -BigInt(i)+y));
            }
        }
    }
    return kopio;
}

function teeKaanto(kaanto, board) {
    if (kaanto == "ylavasen-vasen") {
        return kaannaVasemmalle(BigInt(2),BigInt(2),board);
    } else if (kaanto == "ylavasen-oikea") {
        return kaannaOikealle(BigInt(2),BigInt(2),board);
    } else if (kaanto == "ylaoikea-vasen") {
        return kaannaVasemmalle(BigInt(2),BigInt(5),board);
    } else if (kaanto == "ylaoikea-oikea") {
        return kaannaOikealle(BigInt(2),BigInt(5),board);
    } else if (kaanto == "alavasen-vasen") {
        return kaannaVasemmalle(BigInt(5),BigInt(2),board);
    } else if (kaanto == "alavasen-oikea") {
        return kaannaOikealle(BigInt(5),BigInt(2),board);
    } else if (kaanto == "alaoikea-vasen") {
        return kaannaVasemmalle(BigInt(5),BigInt(5),board);
    } else if (kaanto == "alaoikea-oikea") {
        return kaannaOikealle(BigInt(5),BigInt(5),board);
    }
    return "invalid kaanto";
}

    
function make_whole_move(x, y, board, vuoro) {
    if (vuoro == 1) {
        ans = [board[0], make_move(x,y, board[vuoro])];
        return ans;
    } else {
        ans = [make_move(x,y, board[vuoro]), board[1]]
        return ans;
    }
}
    
function count_set_bits(x) {
    var mask = BigInt(1);
    var ans = 0;
    for (var i = 0; i < 64; i++) {
        if ((mask & x) != 0) ans++;
        mask = mask << BigInt(1);
    }
    return ans;
}    
    
function deduct_turn(board) {
    if (count_set_bits(board[0]) == count_set_bits(board[1])) return 0;
    return 1;
}


function make_whole_turn(kaanto, board) {
    var ans = [teeKaanto(kaanto,board[0]), teeKaanto(kaanto, board[1])];
    return ans;
}

function tee_siirto_ja_kaanto(x, y, kaanto, board) {
    uusboard = make_whole_move(x,y,board, deduct_turn(board));
    ans = [teeKaanto(kaanto,uusboard[0]), teeKaanto(kaanto, uusboard[1])]
    return ans;
}


function convert_to_bitboard(board) {
    bitboard = [BigInt(0), BigInt(0)];
    for (var i = 0; i < 6; i++) {
        for (var j = 0; j < 6; j++) {
            if (board[i][j] == "white") {
                bitboard[0] = make_move(i+1,j+1,bitboard[0]);
            } else if (board[i][j] == "black") {
                bitboard[1] = make_move(i+1,j+1,bitboard[1]);
            }
        }
    }
    return bitboard;
}

var kaannot = ["ylavasen-vasen", "ylavasen-oikea", "ylaoikea-vasen", "ylaoikea-oikea", "alavasen-vasen", "alavasen-oikea", "alaoikea-vasen", "alaoikea-oikea"];

function getMoveDescription(board, white, black) {
    for (var k = 0; k < 8; k++) {
        var kaannetty = make_whole_turn(kaannot[k], board);
        var diff = count_set_bits(kaannetty[0] ^ white) + count_set_bits(kaannetty[1] ^ black);
        if (diff > 1) continue;
        for (var i = 1; i <= 6; i++) {
            for (var j = 1; j <= 6; j++) {
                var result = tee_siirto_ja_kaanto(i,j,kaannot[k],board);
                if (result[0] == white && result[1] == black) {
                    var merkki = String.fromCharCode('a'.charCodeAt() + j - 1);
                    var numero = 7-i;
                    return merkki + numero + " " + kaannot[k];
                }
            }
        }
    }
    return "virhe" + "(" + board[0] + "," + board[1] + ") (" + white + "," + black + ")";
}

function getMoveKaanto(board, white, black) {
    for (var k = 0; k < 8; k++) {
        var kaannetty = make_whole_turn(kaannot[k], board);
        var diff = count_set_bits(kaannetty[0] ^ white) + count_set_bits(kaannetty[1] ^ black);
        if (diff > 1) continue;
        for (var i = 1; i <= 6; i++) {
            for (var j = 1; j <= 6; j++) {
                var result = tee_siirto_ja_kaanto(i,j,kaannot[k],board);
                if (result[0] == white && result[1] == black) {
                    return kaannot[k];
                }
            }
        }
    }
    return "virhe" + "(" + board[0] + "," + board[1] + ") (" + white + "," + black + ")";
}

File: bitboard.py

def laske_mask(x,y):
    return 1 << ((x - 1) * 7 + y - 1)

def make_move(x,y, board):
    #print(x,y,board, laske_mask(x,y))
    return board | laske_mask(x,y)

def loytyy_viiden_suora(x):
    if ((x << 1) & (x >> 1) & (x << 2) & (x >> 2) & x) != 0:
        return True
    if ((x << 7) & (x >> 7) & (x << 14) & (x >> 14) & x) != 0:
        return True
    if ((x << 6) & (x >> 6) & (x << 12) & (x >> 12) & x) != 0:
        return True
    if ((x << 8) & (x >> 8) & (x << 16) & (x >> 16) & x) != 0:
        return True
    return False

def kaannaVasemmalle(x, y, board):
    kopio = board
    for i in range(-1,2):
        for j in range(-1,2):
            bitti = (laske_mask(i+x, j+y) & board)
            if bitti != 0:
                kopio |= laske_mask(-j+x, i+y)
            else:
                kopio &= (~laske_mask(-j+x, i+y))
    return kopio

def kaannaOikealle(x, y, board):
    kopio = board
    for i in range(-1,2):
        for j in range(-1,2):
            bitti = (laske_mask(i+x, j+y) & board)
            if bitti != 0:
                kopio |= laske_mask(j+x, -i+y)
            else:
                kopio &= (~laske_mask(j+x, -i+y))
    return kopio

def teeKaanto(kaanto, board):
    if kaanto == "ylavasen-vasen":
        return kaannaVasemmalle(2,2,board)
    elif kaanto == "ylavasen-oikea":
        return kaannaOikealle(2,2,board)
    elif kaanto == "ylaoikea-vasen":
        return kaannaVasemmalle(2,5, board)
    elif kaanto == "ylaoikea-oikea":
        return kaannaOikealle(2,5, board)
    elif kaanto == "alavasen-vasen":
        return kaannaVasemmalle(5,2, board)
    elif kaanto == "alavasen-oikea":
        return kaannaOikealle(5,2, board)
    elif kaanto == "alaoikea-vasen":
        return kaannaVasemmalle(5,5, board)
    elif kaanto == "alaoikea-oikea":
        return kaannaOikealle(5,5, board)
    return "invalid kaanto"
       

# 0 is white
def symboliKohdassa(x,y, board):
    mask = laske_mask(x,y)
    if mask & board[0] != 0:
        return "0"
    if mask & board[1] != 0:
        return "X"
    return " "

def tulosta(board):
    for i in range(1,7):
        for j in range(1,7):
            print(symboliKohdassa(i,j,board), end ="")
            print("|", end="")
        print("")
    print("")
    
def make_whole_move(x, y, board, vuoro):
    if vuoro == 1:
        return (board[0], make_move(x,y, board[vuoro]))
    else:
        return (make_move(x,y, board[vuoro]), board[1])
    
# 0 = valkoisen vuoro    
def deduct_turn(board):
    if bin(board[0]).count('1') == bin(board[1]).count('1'):
        return 0
    return 1
    
def tee_siirto_ja_kaanto(x,y,kaanto, board):
    #print(deduct_turn(board))
    #print(board)
    board = make_whole_move(x,y,board, deduct_turn(board))
    #print(board)
    return (teeKaanto(kaanto,board[0]), teeKaanto(kaanto, board[1]))

def get_result(board):
    if loytyy_viiden_suora(board[0]) and loytyy_viiden_suora(board[1]):
        return "tasapeli"
    if loytyy_viiden_suora(board[0]):
        return "valkoisen voitto"
    if loytyy_viiden_suora(board[1]):
        return "mustan voitto"
    if bin(board[0]).count('1') + bin(board[1]).count('1') == 36:
        return "tasapeli"
    return "kesken"
    
def testailu():
    board = (0,0)
    vuoro = 0
    while True:
        x = int(input("Anna x\n"))
        y = int(input("Anna y \n"))
        print("Vuoro : " + str(vuoro))
        print(bin(board[vuoro]), bin(make_move(x,y, board[vuoro])))
        if vuoro == 1:
            board = (board[0], make_move(x,y, board[vuoro]))
        else:
            board = (make_move(x,y, board[vuoro]), board[1])
        vuoro = 1 - vuoro
        tulosta(board)
        print(bin(board[0]), bin(board[1]))

#testailu()

tulosta((65536,0))
tulosta((4,1))
tulosta((4,0))

